### PR TITLE
Add the `pre` command

### PR DESCRIPTION
**Pull request by justinmchase**
This adds a new command: `pre`

# Pre Command

```
version pre <subaction>? <identifier>?
```

> **subaction** (optional): `major|minor|patch`
> This will increment using the `pre{subaction}` command

> **identifier** (optional): `alpha|beta|branch`
> The identifier will be prepended to the pre-release version as part of the tag. `alpha` and `beta` are literal identifiers while branch will create a short-hash of the current git branch name.

## Examples

| command          | old              | new              |
| ---------------- | ---------------- | ---------------- |
| pre              | 1.1.0            | 1.1.0-0          |
| pre alpha        | 1.1.0            | 1.1.0-alpha.0    |
| pre major alpha  | 1.1.0            | 2.0.0-alpha.0    |
| pre minor alpha  | 1.1.0            | 1.2.0-alpha.0    |
| pre patch alpha  | 1.1.0            | 1.1.1-alpha.0    |
| pre beta         | 1.1.0            | 1.1.0-beta.0     |
| pre major beta   | 1.1.0            | 2.0.0-beta.0     |
| pre minor beta   | 1.1.0            | 1.2.0-beta.0     |
| pre patch beta   | 1.1.0            | 1.1.1-beta.0     |
| pre branch       | 1.1.0            | 1.1.0-7b30dbf5.0 |
| pre major branch | 1.1.0            | 2.0.0-7b30dbf5.0 |
| pre minor branch | 1.1.0            | 1.2.0-7b30dbf5.0 |
| pre patch branch | 1.1.0            | 1.1.1-7b30dbf5.0 |
| pre              | 1.1.0-alpha.0    | 1.1.0-alpha.1    |
| pre alpha        | 1.1.0-alpha.0    | 1.1.0-alpha.1    |
| pre major alpha  | 1.1.0-alpha.0    | 2.0.0-alpha.0    |
| pre minor alpha  | 1.1.0-alpha.0    | 1.2.0-alpha.0    |
| pre patch alpha  | 1.1.0-alpha.0    | 1.1.1-alpha.0    |
| pre beta         | 1.1.0-alpha.0    | 1.1.0-beta.0     |
| pre branch       | 1.1.0-alpha.0    | 1.1.0-7b30dbf5.0 |
| pre              | 1.1.0-beta.0     | 1.1.0-beta.1     |
| pre beta         | 1.1.0-beta.0     | 1.1.0-beta.1     |
| pre major beta   | 1.1.0-beta.0     | 2.0.0-beta.1     |
| pre minor beta   | 1.1.0-beta.0     | 1.2.0-beta.0     |
| pre patch beta   | 1.1.0-beta.0     | 1.1.1-beta.0     |
| pre alpha        | 1.1.0-beta.0     | 1.1.0-alpha.0    |
| pre branch       | 1.1.0-beta.0     | 1.1.0-branch.0   |
| pre              | 1.1.0-7b30dbf5.0 | 1.1.0-7b30dbf5.1 |
| pre branch       | 1.1.0-7b30dbf5.0 | 1.1.0-7b30dbf5.1 |
| pre major branch | 1.1.0-7b30dbf5.0 | 2.0.0-7b30dbf5.0 |
| pre minor branch | 1.1.0-7b30dbf5.0 | 1.2.0-7b30dbf5.0 |
| pre patch branch | 1.1.0-7b30dbf5.0 | 1.1.1-7b30dbf5.0 |
| pre alpha        | 1.1.0-7b30dbf5.0 | 1.1.0-alpha.0    |
| pre beta         | 1.1.0-7b30dbf5.0 | 1.1.0-beta.0     |

### CODE REVIEW

**Comment by justinmchase** [comment]
I didn't realize you had an issue for this already... great minds think alike!

Fixes https://github.com/dylanpyle/version/issues/14

**Comment by justinmchase** [comment]
This is related to but not blocked by https://github.com/dylanpyle/version/pull/16

Its related insofar as the changes to the `get` command should also include getting prerelease versions with the same logic.

File: hash.ts
Comment (dylanpyle) [inline hash.ts]: I suppose it's probably about time for a `deps.ts` 😄  Think you could move these there?

File: index.ts
Comment (dylanpyle) [inline index.ts]: nit: can you run this through `deno fmt` so we get the right semicolons/etc in place? I'll add a "contributing" section to the readme with this request shortly

**Review by dylanpyle — CHANGES_REQUESTED** [review]
> branch will create a short-hash of the current git branch name.

Just curious, is this a common pattern? Not something I've come across before and I'm a bit hesitant to add a couple new dependencies if it's potentially specific to your setup.

Besides that, looks great after the two tiny nit requests below!

File: index.ts
Comment (justinmchase) [reply index.ts]: derp, yeah

File: hash.ts
Comment (justinmchase) [reply hash.ts]: yup, I considered it.

Maybe an action with a fmt check also? So lazy people like me don't need you to remind me to run fmt.

**Comment by justinmchase** [comment]
Well those two dependencies are very small and from the Deno `std` library so hopefully that's reasonable if the feature is warranted.

But I'll try to walk through my logic around the branch name...

I can't specifically speak to how common it is, though its something I've done in a number of contexts and I've seen a number of times. The reason why its hashed is just that the branch name itself can be in any format and so its not really useable directly in the version identifier. You can use a normalization algorithm, that can get really complex especially if you consider people may put in unicode characters into their branch names... So this is a very generalized normalization technique. I think sometimes others are using the git commit short hash, which is fine but well you can merge a branch and the commit goes to another branch and it can be weird in that case until you add another commit, that's why I like hashing the branch name instead.

But the reason why one may want this _at all_ is to create an artifact from a feature branch before it gets merged to the main line. This will allow upstream projects depending on this to experiment with a particular feature before it gets accepted. Typically someone would just give it a custom name `1.2.0-experimenal-feature-x.0` but its a little hard for us to support that and worse its hard to automate that. With this "branch" feature in this PR, I can essentially support someone easily just publishing a build from any branch in an easy and consistent way. 

I'll sample a few big repos to and see how they're doing:

- [graphql](https://www.npmjs.com/package/graphql?activeTab=versions) - has `alpha`, `beta`, `rc` and feature name identifiers.
- [react](https://www.npmjs.com/package/react?activeTab=versions) - has `rc`, 'beta', 'next', `experimental`, what looks like a commit hash and what looks like a sequential id
- [lerna](https://www.npmjs.com/package/lerna?activeTab=versions) - has 'rc', 'beta', [further back](https://www.npmjs.com/package/lerna/v/2.0.0-beta.38-40-g203f7eb) they were using hashes of some kind too
- [next](https://www.npmjs.com/package/next?activeTab=versions) - just uses `canary` for everything it seems
- [typescript](https://www.npmjs.com/package/typescript?activeTab=versions) - uses `dev` and what appears like the date as an integer, with a possible incrementing id


Tons of repos don't use it at all, especially smaller ones with probably a pretty linear development path. I can tell I should probably ad `rc` to this list though, that seems very common. But I don't think you'd want to support every thing under the sun... probably having the three will guide people to use those instead of coming up with all kinds of different conventions.

I could see some potential additions too like:

 `date` - appends `.20220121` 
 `seq` - appends an incrementing id `.123456` irrespective of the semver changes.

**Pull request closed without merging**
